modules/nsq/src/main/scala/com/snowplowanalytics/snowplow/enrich/nsq/Source.scala

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) 2023-2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.nsq
+
+import cats.syntax.all._
+
+import cats.effect.{Blocker, ConcurrentEffect, ContextShift, Resource, Sync}
+
+import fs2.Stream
+
+import fs2.concurrent.Queue
+
+import org.typelevel.log4cats.Logger
+import org.typelevel.log4cats.slf4j.Slf4jLogger
+
+import com.snowplowanalytics.client.nsq._
+import com.snowplowanalytics.client.nsq.callbacks._
+import com.snowplowanalytics.client.nsq.exceptions.NSQException
+import com.snowplowanalytics.client.nsq.lookup.DefaultNSQLookup
+
+import com.snowplowanalytics.snowplow.enrich.common.fs2.config.io.Input
+
+object Source {
+
+  private implicit def unsafeLogger[F[_]: Sync]: Logger[F] =
+    Slf4jLogger.getLogger[F]
+
+  def init[F[_]: ConcurrentEffect: ContextShift](
+    blocker: Blocker,
+    input: Input
+  ): Stream[F, Record[F]] =
+    input match {
+      case config: Input.Nsq =>
+        for {
+          q <- Stream.eval(Queue.bounded[F, Either[Throwable, Record[F]]](config.maxBufferQueueSize))
+          _ <- Stream.resource(startConsumer(blocker, q, config))
+          msg <- q.dequeueChunk(config.maxBufferQueueSize).flatMap(m => Stream.fromEither[F](m))
+        } yield msg
+      case i =>
+        Stream.raiseError[F](new IllegalArgumentException(s"Input $i is not NSQ"))
+    }
+
+  private def startConsumer[F[_]: Sync: ContextShift: ConcurrentEffect](
+    blocker: Blocker,
+    queue: Queue[F, Either[Throwable, Record[F]]],
+    config: Input.Nsq
+  ): Resource[F, NSQConsumer] =
+    Resource.make(
+      Sync[F].delay {
+        val messageCallback = new NSQMessageCallback {
+          override def message(message: NSQMessage): Unit = {
+            val msgBytes = message.getMessage
+            val enqueue = queue.enqueue1(Right(Record(msgBytes, Sync[F].delay(message.finished()))))
+            ConcurrentEffect[F].toIO(enqueue).unsafeRunSync()
```

Comment:
Are you using `ConcurrentEffect` so that the enqueing can be cancelled or only to get an `IO` that we can run? In the latter case, I think that we can just use `IO` everywhere instead of generic `F`.

Reply:
It is for getting an `IO` to run it. 

I see your point but using `F` still looks better to me. If you aren't strongly opinionated about it, I would vote to leave it as it is.

Reply:
For what it's worth, [this is the equivalent line in fs2-aws](https://github.com/laserdisc-io/fs2-aws/blob/v3.1.1/fs2-aws/src/main/scala/fs2/aws/kinesis/Kinesis.scala#L86).

fs2-google-pubsub does not have an equivalent line, because it uses a `java.util.concurrent.BlockingQueue` instead of a fs2 queue ([here](https://github.com/permutive-engineering/fs2-google-pubsub/blob/v0.18.1/fs2-google-pubsub-grpc/src/main/scala/com/permutive/pubsub/consumer/grpc/internal/PubsubSubscriber.scala#L68)).

When we bump to CE3 then this line will look slightly different.  You would use a `Dispatcher[F]` ([docs](https://typelevel.org/cats-effect/docs/std/dispatcher)) and then:

```
dispatcher.unsafeRunSync(enqueue)
```

Anyway, I don't have a point to make.  I just thought it was interesting.

Reply:
I'm not strongly opinionated and I'm fine with leaving this as it is. There is no great solution anyway to mix functional and non-functional code.